--- a/fs/cifs/smb2ops.c	2018-10-20 10:32:52.000000000 +0000
+++ b/fs/cifs/smb2ops.c	2018-12-20 13:28:00.487325050 +0000
@@ -1043,10 +1043,11 @@
 {
 	struct smb2_hdr *hdr = (struct smb2_hdr *)buf;
 
-	if (hdr->Status != STATUS_NETWORK_SESSION_EXPIRED)
+	if (hdr->Status != STATUS_NETWORK_SESSION_EXPIRED &&
+           hdr->Status != STATUS_USER_SESSION_DELETED)
 		return false;
 
-	cifs_dbg(FYI, "Session expired\n");
+	cifs_dbg(FYI, "Session expired or deleted\n");
 	return true;
 }
 
