--- a/net/core/ethtool.c	2018-12-17 16:05:11.670149817 +0000
+++ b/net/core/ethtool.c	2018-12-17 18:43:02.331772224 +0000
@@ -742,6 +742,16 @@
 	return dev->ethtool_ops->set_link_ksettings(dev, &link_ksettings);
 }
 
+static void
+warn_incomplete_ethtool_legacy_settings_conversion(const char *details)
+{
+       char name[sizeof(current->comm)];
+
+       pr_info_once("warning: `%s' uses legacy ethtool link settings API, %s\n",
+                    get_task_comm(name, current), details);
+}
+
+
 /* Query device for its ethtool_cmd settings.
  *
  * Backward compatibility note: for compatibility with legacy ethtool,
@@ -768,8 +778,10 @@
 							   &link_ksettings);
 		if (err < 0)
 			return err;
-		convert_link_ksettings_to_legacy_settings(&cmd,
-							  &link_ksettings);
+		if(!convert_link_ksettings_to_legacy_settings(&cmd,
+							  &link_ksettings))
+			warn_incomplete_ethtool_legacy_settings_conversion(
+				"link modes are only partially reported");
 
 		/* send a sensible cmd tag back to user */
 		cmd.cmd = ETHTOOL_GSET;
