--- a/fs/cifs/smb2pdu.c	2018-12-11 12:21:15.328265171 +0000
+++ b/fs/cifs/smb2pdu.c	2018-12-11 17:31:31.932403016 +0000
@@ -262,6 +262,9 @@
 	rc = SMB2_tcon(0, tcon->ses, tcon->treeName, tcon, nls_codepage);
 	mutex_unlock(&tcon->ses->session_mutex);
 
+	if (tcon->use_persistent)
+        	cifs_reopen_persistent_handles(tcon);
+
 	cifs_dbg(FYI, "reconnect tcon rc = %d\n", rc);
 	if (rc)
 		goto out;
