--- kernel-exp-4.9/drivers/pci/pci.c	2018-12-17 16:05:12.020144341 +0000
+++ kernel-exp-4.9-new/drivers/pci/pci.c	2018-12-20 12:28:27.422335270 +0000
@@ -3854,6 +3854,9 @@
 {
 	u32 cap;
 
+	if (dev->dev_flags & PCI_DEV_FLAGS_NO_FLR_RESET)
+               return -ENOTTY;
+
 	pcie_capability_read_dword(dev, PCI_EXP_DEVCAP, &cap);
 	if (!(cap & PCI_EXP_DEVCAP_FLR))
 		return -ENOTTY;
--- kernel-exp-4.9/drivers/pci/quirks.c	2018-12-17 16:05:11.711149176 +0000
+++ kernel-exp-4.9-new/drivers/pci/quirks.c	2018-12-20 12:30:52.248062887 +0000
@@ -3348,6 +3348,18 @@
 DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_ATI, PCI_ANY_ID,
 			       PCI_CLASS_DISPLAY_VGA, 8, quirk_no_pm_reset);
 
+static void quirk_no_flr_reset(struct pci_dev *dev)
+{
+       dev->dev_flags |= PCI_DEV_FLAGS_NO_FLR_RESET;
+}
+
+DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_LSI_LOGIC,
+                              PCI_DEVICE_ID_LSI_SAS3008,
+                              PCI_CLASS_STORAGE_SAS, 8, quirk_no_flr_reset);
+DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_LSI_LOGIC,
+                              PCI_DEVICE_ID_LSI_MEGARAID_3108,
+                              PCI_CLASS_STORAGE_SAS, 8, quirk_no_flr_reset);
+
 /*
  * Thunderbolt controllers with broken MSI hotplug signaling:
  * Entire 1st generation (Light Ridge, Eagle Ridge, Light Peak) and part
--- kernel-exp-4.9/include/linux/pci.h	2018-12-17 16:05:12.020144341 +0000
+++ kernel-exp-4.9-new/include/linux/pci.h	2018-12-20 12:33:34.453517805 +0000
@@ -183,6 +183,8 @@
 	 * the direct_complete optimization.
 	 */
 	PCI_DEV_FLAGS_NEEDS_RESUME = (__force pci_dev_flags_t) (1 << 11),
+	/* Do not use FLR reset even if device advertises FLReset+ */
+	PCI_DEV_FLAGS_NO_FLR_RESET = (__force pci_dev_flags_t) (1 << 9),
 };
 
 enum pci_irq_reroute_variant {
--- kernel-exp-4.9/include/linux/pci_ids.h	2018-10-20 10:32:52.000000000 +0000
+++ kernel-exp-4.9-new/include/linux/pci_ids.h	2018-12-20 12:34:43.665431836 +0000
@@ -213,6 +213,8 @@
 #define PCI_DEVICE_ID_LSI_SAS1066E	0x005A
 #define PCI_DEVICE_ID_LSI_SAS1068E	0x0058
 #define PCI_DEVICE_ID_LSI_SAS1078	0x0060
+#define PCI_DEVICE_ID_LSI_SAS3008       0x0097
+#define PCI_DEVICE_ID_LSI_MEGARAID_3108 0x005d
 
 #define PCI_VENDOR_ID_ATI		0x1002
 /* Mach64 */
